### ISSUE TITLE
WalkBuilder: Improved API for adding multiple roots

### ISSUE THREAD

**Issue by ssokolow** [BurntSushi/ripgrep#1761]
In the creations I'm porting from Python or writing anew, the most common use I have for recursive directory traversal is to turn a mixture of file and directory paths given at the command line into a list of file paths to be processed.

While `ignore` and `walkdir` one-up Python's `os.walk` in one respect (I don't need to call `os.path.isdir` and bypass `os.walk` for non-directories), `WalkBuilder` makes taking a `Vec` of roots surprisingly un-ergonomic because the first one must be special-cased.

If I'm willing to mutate it, the simplest I've found is:

```rust
    if let Some(path1) = opts.inpath.pop() {
        let mut builder = WalkBuilder::new(path1);
        for path in opts.inpath {
            builder.add(path);
        }
        for result in builder.build() {
            // TODO: Process each file
        }

    }
```

If I'm not willing to mutate it, then it's worse.

It reminds me of the awkwardness of using `Iterator::fold` when what's really needed is `Iterator::fold_first`.

I'd like to propose one of the following:

* Add a `WalkBuilder::empty()` constructor so I can call `add` for the first argument too.
* Add a `WalkBuilder::from_iter()` constructor so I can just feed my `Vec` of roots directly to it.

**Comment by BurntSushi** [comment]
This sounds fine to me, as long as the semantics for an empty walker are clear and the implementation matches those semantics. (The existing code I imagine has never needed to handle that case, since a walker always has at least one root by construction.) I don't expect this to be a challenge or anything, but it's easy to overlook.

PRs are welcome. I'm unlikely to do this myself any time soon, but could do it next time I'm in the code.

**Comment by ssokolow** [comment]
Fair enough. I certainly won't have time to work on someone else's codebase until at least the new year, but I'll leave myself a note in case I can find time.

Do you have a preference between `WalkBuilder::empty()` and `WalkBuilder::from_iter()` or should I use my own judgment on which one to implement if I get to it first?

**Comment by BurntSushi** [comment]
I like from_iter, but having both doesn't aeem like a terrible idea.

**Comment by SpyrosRoum** [comment]
Hello, I'd like to give this a try, I just have a couple of questions:
1.  What should happen if you use `WalkBuilder::from_iter()` with an empty iterator?
    I assume nothing, it would be the same as using `WalkBuilder::empty()`
2. What should happen if you try to run `WalkBuilder::build()` but you haven't added any paths? Should it panic?

**Comment by kokounet** [comment]
Hello, what's the status of this? I would be interested in this feature as well 😄.

My opinion @SpyrosRoum for building an empty `WalkBuilder` is that it should work fine, but the built `Walk` should just have no items during iteration, that saves us from having to panic.

**Comment by jar-of-salt** [comment]
I am taking a stab at this, opening a PR shortly.

**Comment by jar-of-salt** [comment]
This is complete and ready for review
